### DISCUSSION TITLE
Can I execute code when a query is garbage collected?

### DISCUSSION THREAD

**Discussion by danielbrauer** [TanStack/query#872] [General] [answered]
I'm making an app which decrypts remote image data before displaying it in the browser. The decrypted result is stored using `URL.createObjectURL()`, but this will leak memory if not released at some point using `URL.revokeObjectURL()`. I would love to use React Query to handle fetching and caching this data, but it's not clear to me how I would make sure that the URLs are revoked at the appropriate time.

**Comment by danielbrauer** [comment]
Is this perhaps the sort of thing this PR is intended to enable? https://github.com/tannerlinsley/react-query/pull/867

**Comment by pianomansam** [comment, marked as answer]
Five years later, but this is how I just accomplished this very thing in v5:

```
queryClient.getQueryCache().subscribe((event) => {
  if (
    event.type === 'removed' &&
    ['CanvasCroppedImage', 'CanvasMultiTagImage'].includes(
      event.query.queryKey[0],
    ) &&
    typeof event.query.state.data === 'string' &&
    event.query.state.data.length > 0
  ) {
    URL.revokeObjectURL(event.query.state.data);
  }
});
```

**Reply by TkDodo** [reply]
That’s not bad 👍 . You can also use `meta` to “tag” queries that are url images.
